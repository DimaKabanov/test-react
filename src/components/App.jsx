import React from 'react';
import Tabs from './Tabs';

const App = (props) => {
  const { tabs } = props;

  return (
    <div>
      <h1>My first React App</h1>
      <Tabs tabs={tabs} />
    </div>
  );
};

export default App;
