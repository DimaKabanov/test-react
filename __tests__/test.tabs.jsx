import '@babel/polyfill';
import { promises as fs } from 'fs';
import nock from 'nock';
import delay from 'delay';
import 'react-log-state';
import App from '../src/components/App';
import parseXml from '../src/utils/parser';

ReactLogState.logAll();

const tabControlsSelector = 'li[data-test="tab-control"]';
const addNewTabBtnSelector = '[data-test="add-tab-btn"]';
const removeTabBtnSelector = '[data-test="remove-tab-btn"]';
const tabControlsContainerSelector = 'ul[data-test="tab-controls-container"]';
const rssInputSelector = '[data-test="rss-link-input"]';
const rssFormSelector = '[data-test="form-rss"]';

const getSelector = app => ({
  getTabControlList: () => app.find(tabControlsSelector),
  getTabControsContainer: () => app.find(tabControlsContainerSelector),
  getTabByIndex: index => app.find(tabControlsSelector).at(index),
  getAddNewTabBtn: () => app.find(addNewTabBtnSelector),
  getRemoveTabBtn: () => app.find(removeTabBtnSelector),
  getRssInput: () => app.find(rssInputSelector),
  getRssForm: () => app.find(rssFormSelector),
});

describe('Tabs snapshot', () => {
  test('Renders', () => {
    const app = mount(<App />);
    expect(app.render()).toMatchSnapshot();
  });

  test('Click to second tab', () => {
    const app = mount(<App />);
    const s = getSelector(app);
    const secondTab = s.getTabByIndex(1);

    secondTab.simulate('click');
    expect(app.render()).toMatchSnapshot();
  });
});

describe('Tabs without snapshot', () => {
  test('Correct switching to second tab', () => {
    const app = mount(<App />);
    const s = getSelector(app);
    const secondTab = s.getTabByIndex(1);

    secondTab.simulate('click');

    const firstTab = s.getTabByIndex(0);
    const updatedSecondTab = s.getTabByIndex(1);

    expect(firstTab).toHaveProp('aria-selected', 'false');
    expect(updatedSecondTab).toHaveProp('aria-selected', 'true');
  });

  test('Correct adding new tab', () => {
    const app = mount(<App />);
    const s = getSelector(app);
    const addNewTabBtn = s.getAddNewTabBtn();

    addNewTabBtn.simulate('click');

    const tabControlsContainer = s.getTabControsContainer();
    expect(tabControlsContainer).toContainMatchingElements(4, tabControlsSelector);
  });

  test('Correct removal tab', () => {
    const app = mount(<App />);
    const s = getSelector(app);
    const removeTabBtn = s.getRemoveTabBtn();

    removeTabBtn.simulate('click');

    const tabControlsContainer = s.getTabControsContainer();
    expect(tabControlsContainer).toContainMatchingElements(2, tabControlsSelector);
  });

  test('Correct set active tab after reload', () => {
    const cookiesStub = () => {
      const cookies = {};
      return {
        set: (name, value) => { cookies[name] = value; },
        get: name => cookies[name],
      };
    };
    const cookies = cookiesStub();
    const app = mount(<App cookies={cookies} />);
    const s = getSelector(app);
    const secondTab = s.getTabByIndex(1);
    expect(secondTab).toHaveProp('aria-selected', 'false');

    secondTab.simulate('click');

    const app2 = mount(<App cookies={cookies} />);
    const s2 = getSelector(app2);

    const updatedSecondTab = s2.getTabByIndex(1);
    expect(updatedSecondTab).toHaveProp('aria-selected', 'true');
  });

  test('Correct set RSS data', async () => {
    const testXmlPath = '__tests__/__fixtures__/rss.xml';
    const xml = await fs.readFile(testXmlPath, { encoding: 'utf8' });
    const { title } = parseXml(xml);
    const host = 'https://cors-anywhere.herokuapp.com/';
    const url = 'test';
    nock(host)
      .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
      .get(`/${url}`)
      .reply(200, xml);

    const app = mount(<App />);
    const s = getSelector(app);
    const tabControlsContainer = s.getTabControsContainer();
    expect(tabControlsContainer).toContainMatchingElements(3, tabControlsSelector);

    const rssInput = s.getRssInput();
    rssInput.simulate('change', { target: { value: url } });
    const rssForm = s.getRssForm();
    rssForm.simulate('submit');
    await delay(100);
    app.update();

    const tabControlsContainerAfterSave = s.getTabControsContainer();
    expect(tabControlsContainerAfterSave).toContainMatchingElements(4, tabControlsSelector);
    expect(tabControlsContainerAfterSave).toIncludeText(title);
  });
});
