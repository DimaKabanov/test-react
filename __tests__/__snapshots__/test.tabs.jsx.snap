// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs snapshot Click to second tab 1`] = `
<div>
  <h1>
    My first React App
  </h1>
  <div
    class="react-tabs"
    data-tabs="true"
  >
    <button
      data-test="add-tab-btn"
      type="button"
    >
      Add tab
    </button>
    <form
      data-test="form-rss"
    >
      <label
        for="rss-link"
      >
        RSS link
        <input
          data-test="rss-link-input"
          id="rss-link"
          type="text"
        />
      </label>
      <button
        type="submit"
      >
        Add RSS
      </button>
    </form>
    <ul
      class="react-tabs__tab-list"
      data-test="tab-controls-container"
      role="tablist"
    >
      <li
        aria-controls="react-tabs-7"
        aria-disabled="false"
        aria-selected="false"
        class="react-tabs__tab"
        data-test="tab-control"
        id="react-tabs-6"
        role="tab"
      >
        Tab 1
      </li>
      <li
        aria-controls="react-tabs-9"
        aria-disabled="false"
        aria-selected="true"
        class="react-tabs__tab react-tabs__tab--selected"
        data-test="tab-control"
        id="react-tabs-8"
        role="tab"
        tabindex="0"
      >
        Tab 2
      </li>
      <li
        aria-controls="react-tabs-11"
        aria-disabled="false"
        aria-selected="false"
        class="react-tabs__tab"
        data-test="tab-control"
        id="react-tabs-10"
        role="tab"
      >
        Tab 3
      </li>
    </ul>
    <div
      aria-labelledby="react-tabs-6"
      class="react-tabs__tab-panel"
      id="react-tabs-7"
      role="tabpanel"
    />
    <div
      aria-labelledby="react-tabs-8"
      class="react-tabs__tab-panel react-tabs__tab-panel--selected"
      id="react-tabs-9"
      role="tabpanel"
    >
      Tab text 2
      <button
        data-test="remove-tab-btn"
        type="button"
      >
        Remove tab 5
      </button>
    </div>
    <div
      aria-labelledby="react-tabs-10"
      class="react-tabs__tab-panel"
      id="react-tabs-11"
      role="tabpanel"
    />
  </div>
</div>
`;

exports[`Tabs snapshot Renders 1`] = `
<div>
  <h1>
    My first React App
  </h1>
  <div
    class="react-tabs"
    data-tabs="true"
  >
    <button
      data-test="add-tab-btn"
      type="button"
    >
      Add tab
    </button>
    <form
      data-test="form-rss"
    >
      <label
        for="rss-link"
      >
        RSS link
        <input
          data-test="rss-link-input"
          id="rss-link"
          type="text"
        />
      </label>
      <button
        type="submit"
      >
        Add RSS
      </button>
    </form>
    <ul
      class="react-tabs__tab-list"
      data-test="tab-controls-container"
      role="tablist"
    >
      <li
        aria-controls="react-tabs-1"
        aria-disabled="false"
        aria-selected="true"
        class="react-tabs__tab react-tabs__tab--selected"
        data-test="tab-control"
        id="react-tabs-0"
        role="tab"
        tabindex="0"
      >
        Tab 1
      </li>
      <li
        aria-controls="react-tabs-3"
        aria-disabled="false"
        aria-selected="false"
        class="react-tabs__tab"
        data-test="tab-control"
        id="react-tabs-2"
        role="tab"
      >
        Tab 2
      </li>
      <li
        aria-controls="react-tabs-5"
        aria-disabled="false"
        aria-selected="false"
        class="react-tabs__tab"
        data-test="tab-control"
        id="react-tabs-4"
        role="tab"
      >
        Tab 3
      </li>
    </ul>
    <div
      aria-labelledby="react-tabs-0"
      class="react-tabs__tab-panel react-tabs__tab-panel--selected"
      id="react-tabs-1"
      role="tabpanel"
    >
      Tab text 1
      <button
        data-test="remove-tab-btn"
        type="button"
      >
        Remove tab 1
      </button>
    </div>
    <div
      aria-labelledby="react-tabs-2"
      class="react-tabs__tab-panel"
      id="react-tabs-3"
      role="tabpanel"
    />
    <div
      aria-labelledby="react-tabs-4"
      class="react-tabs__tab-panel"
      id="react-tabs-5"
      role="tabpanel"
    />
  </div>
</div>
`;
